From 501067490d6316d2fe04058deaa4f3d8f6a62250 Mon Sep 17 00:00:00 2001
From: Harshesh Valera <harshesh.v@variscite.com>
Date: Tue, 4 Oct 2022 19:52:12 -0700
Subject: [PATCH] imx8m soms: Update firmware loading options

- Move vendor specific file to vendor image as this was violating
the sepolicies

- Use kernel command line properties and read sdio id to set proper
firmware

- Use single conf file to load LWB5 and LWB firmware

Signed-off-by: Harshesh Valera <harshesh.v@variscite.com>
Change-Id: I1a6a9de5d06fa771bfea559b303181741ebe2bc5
---
 include/bt_vendor_brcm.h |  2 +-
 src/bt_vendor_brcm.c     | 36 ++++++++++++++++++++++++++++++++++++
 src/conf.c               |  3 +++
 3 files changed, 40 insertions(+), 1 deletion(-)

diff --git a/include/bt_vendor_brcm.h b/include/bt_vendor_brcm.h
index bf08b3a..5e6b67f 100644
--- a/include/bt_vendor_brcm.h
+++ b/include/bt_vendor_brcm.h
@@ -51,7 +51,7 @@
 
 /* Run-time configuration file */
 #ifndef VENDOR_LIB_CONF_FILE
-#define VENDOR_LIB_CONF_FILE "/etc/bluetooth/bt_vendor.conf"
+#define VENDOR_LIB_CONF_FILE "/vendor/etc/bluetooth/bt_vendor.conf"
 #endif
 
 /* Device port name where Bluetooth controller attached */
diff --git a/src/bt_vendor_brcm.c b/src/bt_vendor_brcm.c
index 8f4eeeb..d141523 100644
--- a/src/bt_vendor_brcm.c
+++ b/src/bt_vendor_brcm.c
@@ -100,6 +100,42 @@ static const tUSERIAL_CFG userial_init_cfg =
 **
 *****************************************************************************/
 
+int hw_set_firmware_prop(char *p_conf_name, char *p_conf_value, int param)
+{
+	char fw_path_value[PROPERTY_VALUE_MAX];
+	char buf[CONF_MAX_LINE_LEN];
+	int fd = -1;
+	int sz;
+	property_get("ro.boot.bt_sdio", fw_path_value, "0");
+	if ((strcmp(fw_path_value, "0") != 0))
+	{
+		fd = open(fw_path_value, O_RDONLY);
+		if (fd < 0)
+		{
+			ALOGE("failed to read from %s:(%d)\n",
+				fw_path_value, errno);
+			return -1;
+		}
+		else
+		{
+			sz = read(fd, &buf, sizeof(buf));
+			if (strncasecmp(buf, "0x4339", strlen("0x4339")) == 0)
+			{
+				if (strncasecmp(p_conf_name, "BT_FIRMWARE_LWB5", 
+					strlen("BT_FIRMWARE_LWB5")) == 0)
+					property_set("ro.boot.bt_firmware", p_conf_value);
+			}
+			else 
+			{ 
+				if (strncasecmp(p_conf_name, "BT_FIRMWARE", 
+					strlen("BT_FIRMWARE")) == 0)
+					property_set("ro.boot.bt_firmware", p_conf_value);
+			}
+		}
+		close(fd);
+	}
+	return 0;
+}
 static int init(const bt_vendor_callbacks_t* p_cb, unsigned char *local_bdaddr)
 {
     char uart_value[PROPERTY_VALUE_MAX];
diff --git a/src/conf.c b/src/conf.c
index c28c4f1..77eddd1 100644
--- a/src/conf.c
+++ b/src/conf.c
@@ -38,6 +38,7 @@
 int userial_set_port(char *p_conf_name, char *p_conf_value, int param);
 int hw_set_patch_file_path(char *p_conf_name, char *p_conf_value, int param);
 int hw_set_patch_file_name(char *p_conf_name, char *p_conf_value, int param);
+int hw_set_firmware_prop(char *p_conf_name, char *p_conf_value, int param);
 #if (VENDOR_LIB_RUNTIME_TUNING_ENABLED == TRUE)
 int hw_set_patch_settlement_delay(char *p_conf_name, char *p_conf_value, int param);
 #endif
@@ -71,6 +72,8 @@ static const conf_entry_t conf_table[] = {
     {"UartPort", userial_set_port, 0},
     {"FwPatchFilePath", hw_set_patch_file_path, 0},
     {"FwPatchFileName", hw_set_patch_file_name, 0},
+    {"BT_FIRMWARE_LWB5", hw_set_firmware_prop, 0},
+    {"BT_FIRMWARE_LWB", hw_set_firmware_prop, 0},
 #if (VENDOR_LIB_RUNTIME_TUNING_ENABLED == TRUE)
     {"FwPatchSettlementDelay", hw_set_patch_settlement_delay, 0},
 #endif
-- 
2.25.1

